code/controllers/subsystem/mapping.dm

```
@@ -0,0 +1,801 @@
+
+	// Get sector boundaries to calculate relative positions
+	var/list/bounds = get_sector_bounds(alloc.sector)
+	var/x_min = bounds["x_min"]
+	var/y_min = bounds["y_min"]
+
+	// Create matrix with relative coordinates
+	var/datum/turf_matrix[SECTOR_SIZE][SECTOR_SIZE]
+	for (var/turf/T in search_turfs)
+		var/rel_x = T.x - x_min + 1
+		var/rel_y = T.y - y_min + 1
+		turf_matrix[rel_x][rel_y] = T
+
+	// Define safe zone boundaries (accounting for edge buffer and shuttle size)
+	var/safe_x_min = LANDING_ZONE_EDGE_BUFFER + 1
+	var/safe_x_max = SECTOR_SIZE - LANDING_ZONE_EDGE_BUFFER - x_dim
+	var/safe_y_min = LANDING_ZONE_EDGE_BUFFER + 1
+	var/safe_y_max = SECTOR_SIZE - LANDING_ZONE_EDGE_BUFFER - y_dim
+
+	if(safe_x_max < safe_x_min || safe_y_max < safe_y_min)
+		return null // Not enough space for safe landing
+
+	// Create randomized search list within safe boundaries
+	var/list/search_positions = list()
+	for(var/rel_x = safe_x_min; rel_x <= safe_x_max; rel_x++)
+		for(var/rel_y = safe_y_min; rel_y <= safe_y_max; rel_y++)
+			var/turf/T = turf_matrix[rel_x][rel_y]
+			if(T && !iswall(T) && !istype(T, /turf/unsimulated/mineral))
+				search_positions += T
+
+	// Shuffle the search positions for randomization
+	if(!search_positions.len)
+		return null
+
+	search_positions = shuffle(search_positions)
+
+	// Search through randomized positions
+	for(var/turf/T in search_positions)
+		var/rel_x = T.x - x_min + 1
+		var/rel_y = T.y - y_min + 1
+		var/found = TRUE
+
+		for (var/dx = 0; dx < x_dim && found; dx++)
+			for (var/dy = 0; dy < y_dim && found; dy++)
+				var/check_x = rel_x + dx
+				var/check_y = rel_y + dy
+				if(check_x > SECTOR_SIZE || check_y > SECTOR_SIZE) // Out of sector bounds
+					found = FALSE
+					continue
+				var/turf/target = turf_matrix[check_x][check_y]
+				if (!target || !istype(target,T.type))
+					found = FALSE
+
+		if (found)
+			return T  // Return top-left turf of matching rectangle
+
+	return null
```

That sounds like it should throw an error or something? Or some way to get feedback at the user-level? At the very least it should distinguish between "your shuttle is too fat to land here" and "there is no space at all"

---

This proc and its caller don't have a user var; travel_to_planet() does and provides the error message on line 356 when this happens:

to_chat(user, "<span class='warning'>No suitable landing zone found on [planet.planet_name].</span>")